### ISSUE TITLE
useQueryReducer and useLocationReducer

### ISSUE THREAD

**Issue by us88** [xiel/use-location-state#120]
Any chance an implementation of this can be done?
It would be nice to have a `useReducer` variant for query strings.

My use case is filters on a product listing page. Filters exhibit non-trivial behavior in my case, and I was using `useReducer` before I wanted the query string to be the source of truth, and stumbled upon this nice little library.

**Comment by xiel** [comment]
Hey @Magtheridon96 

thats a very good suggestion, shouldn’t be too much work add it. 💡

Are you able to share your existing reducer function and action types as inspiration? 
Might be nice to have a real world example, otherwise also fine :)

May I ask which version / router you use?

– Felix

**Comment by Magtheridon96** [comment]
Hello!

We're using React 16.11, Router 4.3.1, and the reducer looks like this:

```javascript
const filterKeyReducer = (filterKeys, action) => {
  switch (action.type) {
    case 'clear':
      return [];
    case 'add':
      return [...new Set(filterKeys.concat(action.toAdd).filter(x => x))];
    case 'remove':
      return filterKeys.filter(x => !action.toRemove.includes(x));
    /* Required for atomic deletions & additions */
    case 'remove-then-add':
      return [...new Set(
        filterKeys
          .filter(x => !action.toRemove.includes(x))
          .concat(action.toAdd)
          .filter(x => x))];
    default:
      throw new Error('Unhandled action in filterKeyReducer: ' + action.type);
  }
};
```

The code using it to achieve our desired semantics is pretty bulky, but a key excerpt is the logic for handling accordions:
```javascript
...
if (isAccordionOpen) {
  if (noDescendantsIncluded) {
    dispatchFilterKeys({ type: 'remove', toRemove: [key] });
  } else {
    dispatchFilterKeys({ type: 'remove-then-add', toRemove: descendantKeys, toAdd: [key] });
  }
} else {
  // When we are opening an accordion, all nodes in the tree need to be purged.
  // It is not enough to remove siblings, because we might have a sibling's child currently included.
  dispatchFilterKeys({ type: 'remove-then-add', toRemove: allKeysInTree, toAdd: [key] });
}
...
```

When I tried to make the query string a source of truth, I started out with a hacky solution that calls `history.push` inside the reducer with a 0-second timer, but this wasn't robust enough as it still needed me to include logic to sync the state with the query string when the query string changes (and also, it would break correctness if the reducer had side effects I suppose).

Googling "hooks query params" got me here thankfully!

**Comment by xiel** [comment]
@Magtheridon96 Thanks for providing the reducer & example! 

I'll try to get back to you soon as possible (might take me a few days).

**Comment by xiel** [comment]
Hey @Magtheridon96 ,

good news: I'm done with implementing the feature. It is released under the **next** tag for now (**2.4.0**).

```sh
yarn add react-router-use-location-state@next
```

Here is also a demo [sandbox](https://codesandbox.io/s/github/xiel/use-location-state/tree/master/src/examples/react-router-use-location-state?file=/src/pages/QueryReducer/QueryReducer.tsx) (under **useQueryReducer**): 
https://vzjex.csb.app/query-reducer?filters=shirts&filters=jackets&filters=shorts&filters=boots&filters=running

Please give it a try and let be know if you run into problems or if it works as expected.
– Felix
